src/UniswapV4Migrator.sol

```diff
@@ -0,0 +1,178 @@
+
+        return address(0);
+    }
+
+    function migrate(
+        uint160 sqrtPriceX96,
+        address token0,
+        address token1,
+        address recipient
+    ) external payable onlyAirlock returns (uint256 liquidity) {
+        AssetData memory assetData = getAssetData[token0][token1];
+        PoolKey memory poolKey = assetData.poolKey;
+
+        // Let's check if the pool was initialized
+        (uint160 currentSqrtPriceX96,,,) = poolManager.getSlot0(poolKey.toId());
+
+        if (currentSqrtPriceX96 == 0) {
+            IPoolManager(poolManager).initialize(poolKey, sqrtPriceX96);
+        }
+
+        uint256 balance1 = ERC20(token1).balanceOf(address(this));
+
+        uint256 balance0;
+        bytes memory actions;
+        bytes[] memory params;
+
+        if (token0 == address(0)) {
+            params = new bytes[](4);
+            params[3] = abi.encode(CurrencyLibrary.ADDRESS_ZERO, address(this));
+            balance0 = address(this).balance;
+            actions = abi.encodePacked(
+                uint8(Actions.MINT_POSITION),
+                uint8(Actions.MINT_POSITION),
+                uint8(Actions.SETTLE_PAIR),
+                uint8(Actions.SWEEP)
+            );
+        } else {
+            params = new bytes[](3);
+            balance0 = ERC20(token0).balanceOf(address(this));
+            actions =
+                abi.encodePacked(uint8(Actions.MINT_POSITION), uint8(Actions.MINT_POSITION), uint8(Actions.SETTLE_PAIR));
+        }
+
+        int24 lowerTick =
+            ((TickMath.MIN_TICK + 1) - poolKey.tickSpacing + 1) / poolKey.tickSpacing * poolKey.tickSpacing;
+        int24 upperTick = (TickMath.MAX_TICK - 1) / poolKey.tickSpacing * poolKey.tickSpacing;
+
+        int24 currentTick = TickMath.getTickAtSqrtPrice(sqrtPriceX96);
+        require(currentTick >= lowerTick && currentTick <= upperTick, "UniswapV4Migrator: TICK_OUT_OF_RANGE");
+
+        liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtPriceAtTick(lowerTick),
+            TickMath.getSqrtPriceAtTick(upperTick),
+            uint128(balance0),
+            uint128(balance1)
+        );
+
+        require(liquidity > 0, "UniswapV4Migrator: ZERO_LIQUIDITY");
+
+        uint256 lockedLiquidity = liquidity / 10;
+        uint256 timeLockLiquidity = liquidity - lockedLiquidity;
+
+        // Liquidity for the protocol locker
+        params[0] = abi.encode(
+            poolKey,
+            lowerTick,
+            upperTick,
+            uint128(lockedLiquidity),
+            uint128(balance0) / 10,
+            uint128(balance1) / 10,
+            address(this),
+            new bytes(0)
+        );
+
+        // Liquidity for the Timelock
+        params[1] = abi.encode(
+            poolKey,
+            lowerTick,
+            upperTick,
+            uint128(timeLockLiquidity),
+            uint128(balance0),
+            uint128(balance1),
```

Comment:
I think we need to make these be 90% of the balances right?

Reply:
ah nvm this should already be covered with the calculation for `timeLockLiquidity`